app.py

```diff
@@ -0,0 +1,137 @@
+app = Flask(__name__)
+
+
+def fetch_user_data(net_id: str) -> Optional[Dict[str, str]]:
+    """
+    Exports a REDCap record matching the given *net_id*. Returns None if no
+    match is found.
+
+    Raises an :class:`AssertionError` if REDCap returns multiple matches for the
+    given *net_id*.
+    """
+    filter_logic = f'[netid] = "{net_id}"'
+    data = {
+        'token': REDCAP_API_TOKEN,
+        'content': 'record',
+        'format': 'json',
+        'type': 'flat',
+        'csvDelimiter': '',
+        'fields[0]': 'netid',
+        'fields[1]': 'record_id',
+        'fields[2]': 'eligibility_screening_complete',
+        'fields[3]': 'consent_form_complete',
+        'fields[4]': 'enrollment_questionnaire_complete',
+        'filterLogic': filter_logic,
+        'rawOrLabel': 'raw',
+        'rawOrLabelHeaders': 'raw',
+        'exportCheckboxLabel': 'false',
+        'exportSurveyFields': 'false',
+        'exportDataAccessGroups': 'false',
+        'returnFormat': 'json'
+    }
+    response = requests.post(REDCAP_API_URL, data=data)
+    response.raise_for_status()
+
+    if len(response.json()) == 0:
+        return None
+
+    assert len(response.json()) == 1, \
+        f"Error: Multiple records matching {filter_logic}"
+
+    return response.json()[0]
+
+def register_net_id(net_id: str) -> str:
+    """
+    Returns the REDCap record ID of the participant newly registered under the
+    given *net_id*.
+    """
+    # REDCap enforces that we must provide a non-empty record ID. Because we're
+    # using `forceAutoNumber` in the POST request, we do not need to provide a
+    # real record ID.
+    values = [{'netid': net_id, 'record_id': 'record ID cannot be blank'}]
+    data = {
+        'token': REDCAP_API_TOKEN,
+        'content': 'record',
+        'format': 'json',
+        'type': 'flat',
+        'overwriteBehavior': 'normal',
+        'forceAutoNumber': 'true',
+        'data': json.dumps(values),
+        'returnContent': 'ids',
+        'returnFormat': 'json'
+    }
+    response = requests.post(REDCAP_API_URL, data=data)
+    response.raise_for_status()
+    return response.json()[0]
+
+
+def generate_survey_link(record_id: str) -> str:
+    """
+    Returns a generated survey link to the eligibility screening instrument for
+    the given *record_id*.
+    """
+    data = {
+        'token': REDCAP_API_TOKEN,
+        'content': 'surveyLink',
+        'format': 'json',
+        'instrument': 'eligibility_screening',
+        'event': 'enrollment_arm_1',  # TODO subject to change
+        'record': record_id,
+        'returnFormat': 'json'
+    }
+    response = requests.post(REDCAP_API_URL, data=data)
+    response.raise_for_status()
+    return response.text
+
+
+@app.route('/')
+def main():
+    # Get NetID from Shibboleth data
+    net_id = request.remote_user
+
+    while not net_id:
+        # TODO: Redirect to Shibboleth login
+        net_id = 'KaasenG'
+        pass
```

Comment:
Why did you use a `while` here? I would have expected an `if`.

Hmm, and we'll need an `else` clause to handle the case of a missing NetID (which should never happen, unless we misconfigure auth).

(Also, what is `KaasenG`? :-)

Reply:
I wasn't sure how the Shibboleth interaction would look. I used the `while` statement in case there's some missing NetID on the first attempt, and so I would want to send users back to Shibboleth.

Basically I expected this block of code to change quite a bit.

> `KaasenG`

I thought you would never ask! It's a fake NetID for Gunnar Kaasen, [America's favorite husky musher](https://en.wikipedia.org/wiki/Gunnar_Kaasen)!

Reply:
Ah, nice! I googled a little but didn't split up the G and Kaasen.

The Shibboleth interaction will happen entirely outside of this Flask app. So the only check we need to do here right now is check if `request.remote_user` has a value or not.